src/builtin.c

```diff
@@ -523,6 +523,89 @@ static jv f_utf8bytelength(jq_state *jq, jv input) {
   return jv_number(jv_string_length_bytes(input));
 }
 
+# include <utf8proc.h>
+
+static utf8proc_int32_t utf8proc_codepoint_func(utf8proc_int32_t codepoint, utf8proc_int32_t (*func)(utf8proc_int32_t)) {
+  return func(codepoint);
+}
+
+static unsigned char *utf8_func(const unsigned char *str, size_t *len, utf8proc_int32_t (*func)(utf8proc_int32_t)) {
+  utf8proc_uint8_t *output;
+  /* options = UTF8PROC_COMPOSE | UTF8PROC_COMPAT
+  | UTF8PROC_CASEFOLD | UTF8PROC_IGNORE | UTF8PROC_STRIPMARK | UTF8PROC_STRIPCC | UTF8PROC_STRIPNA;
+  */
+  utf8proc_option_t options = UTF8PROC_STRIPNA;
+
+  // utf8proc_map_custom allocates new mem
+  *len = (size_t) utf8proc_map_custom((const utf8proc_uint8_t *)str, *len, &output,
+                                      options,
+                                      (utf8proc_custom_func)utf8proc_codepoint_func, (void *)func);
+  return (unsigned char *)output;
+}
+
+static unsigned char *utf8_toCase(const unsigned char *str, size_t *lenp, int *was_malformed, utf8proc_int32_t (*func)(utf8proc_int32_t)) {
+  size_t len_orig = *lenp;
+  unsigned char *new_s = utf8_func(str, lenp, func);
+  if(!new_s && len_orig) { // malformed
+    unsigned char *tmp_s = malloc(len_orig + 1);
+    if(tmp_s) {
+      *was_malformed = 1;
+      memcpy(tmp_s, str, len_orig);
+      tmp_s[len_orig] = '\0';
+      new_s = utf8_func(tmp_s, lenp, func);
+      free(tmp_s);
+    }
+  }
+  return new_s;
+}
+
+/**
+ * utf8_toLowerCase, utf8_toUpperCase
+ * @param str  string to convert
+ * @param lenp returns length of newly-allocated string
+ * @param was_malformed returns 1 if the input data was malformed
+ * @return newly-allocated lower-case string (caller must free)
+ */
+static unsigned char *utf8_toLowerCase(const unsigned char *str, size_t *lenp, int *was_malformed) {
+  return utf8_toCase(str, lenp, was_malformed, utf8proc_tolower);
+}
+
+static unsigned char *utf8_toUpperCase(const unsigned char *str, size_t *lenp, int *was_malformed) {
+  return utf8_toCase(str, lenp, was_malformed, utf8proc_toupper);
+}
+
+static jv f_utf8tocase(jq_state *jq, jv input, char toLower) {
+  unsigned char *(*changeCase)(const unsigned char *, size_t *, int *) =
+    toLower ? utf8_toLowerCase : utf8_toUpperCase;
+  if(jv_get_kind(input) != JV_KIND_STRING) {
+    return input;
+  } else {
+    const char *jstr = jv_string_value(input);
+    size_t len = jv_string_length_bytes(jv_copy(input));
+    if(len) {
+      int malformed = 0;
+      unsigned char *changed = changeCase((const unsigned char *)jstr, &len, &malformed);
+      if(malformed || !len)
+        free(changed);
```

Comment:
shouldn't we report an error here? or a warning?

Reply:
I don't think so. A better and more flexible alternative would be to add two built-ins along the lines of `utf8_valid` (return true iff input is valid utf8) and `utf8_replace_malformed` (replace any malformed utf8 in some specified manner). In any case it seems to me this issue is more related to https://github.com/jqlang/jq/issues/2660 than to these particular `upcase` and `downcase` functions.

Reply:
That said I don't really feel strongly and defer to whatever you maintainers decide

Reply:
@leonid-s-usov, @itchyny, @nicowilliams, @emanuele6 - I'm inclined to think that it's too much to expect @liquidaty to resolve the larger issue (#2660) here, and would therefore propose that the particular issue regarding `malformed` not stand in the way of this PR being accepted, though I think it would be appropriate to add documentation to the manual regarding the resolution of the anomaly, even if it's only to say the behavior is "undefined".

While on the subject of handling anomalies - is there a precedent that should be followed regarding "warning" messages on stderr?

Reply:
ack